otherlibs/str/str.mli

```diff
@@ -18,6 +18,11 @@
 
 (** {1 Regular expressions} *)
 
+(**
+   The regexp library implements regular expressions on strings interpreted
+   as sequences of 8-bit characters. If you need to handle regular expressions
+   of unicode (extended) grapheme clusters, this library is not a good fit.
```

Comment:
I don't know what a "grapheme cluster" is, let alone an "extended grapheme cluster".  Could we find a simpler terminology?

Reply:
Does `user-perceived character` (or `human-perceived character`) read better?

Reply:
I think that some level of redundancy is useful here for people which are not that certain of what "unicode-aware" means. Maybe
```
If you need to handle unicode-aware regular expressions of user-perceived characters, ...
```

Reply:
> ...matching will be performed at the level of the UTF-8 encoding of Unicode grapheme
clusters...

Not correct. I would propose:

> The Str library implements regular expression matching on strings interpreted as sequences of octets (8-bit character codes). It does not handle Unicode regular expressions or matching on Unicode texts, except those encoded with UTF-8 and comprising only characters from the ASCII character set.

Reply:
We got rid of the graphemes!  Great!

Reply:
An ascii regular expression can work and give the expected result on an Unicode encoded text, even if it contains non-ascii scalar values. Consequently, I think it is better to focus on the regular expression itself:

> The `Str` library does not provide unicode-aware regular expressions. The `Str` library implements regular expression matching on strings interpreted as sequences of Char.t (8-bit character codes) whereas unicode texts use different notions of user-perceived characters and character classes. The two notions of regular expressions are incompatible.

Reply:
> An ascii regular expression can work and give the expected result on an Unicode encoded text, even if it contains non-ascii scalar values. 

It can also falsely match ASCII sequences that are actually parts of multi-octet UTF-8 encoded code points. Which means that an ASCII regular expression does not *really* work unless the text is encoded in the ASCII subset of UTF-8.

Reply:
> It can also falsely match ASCII sequences that are actually parts of multi-octet UTF-8 encoded code points.

There is no such thing: UTF-8 is a prefix code. See the [table here](https://www.rfc-editor.org/rfc/rfc3629#section-3) to convince yourself.

My own take on it would be: 

```
The `Str` library provides regular expressions on sequences of 
bytes. It is, in general, unsuitable to match Unicode characters. 
```

This will deter and avoid surprises for people who have little knowledge of Unicode and be perfectly understandable for people who are more subtle in their understanding and know how to hack their way whenever possible.

Reply:
> There is no such thing: UTF-8 is a prefix code

I think the intended example is something like the regexp `a..b` matching the UTF-8 encoding of `aéb`, which can be surprising indeed, even though the regexp is pure ASCII.

Reply:
I like @dbuenzli 's version: it is short, factual and understandable by a wide audience.

Reply:
Once again @dbuenzli for the win.

Reply:
> I forgot how UTF-8 encodes, and I wrote a codec for it. Doh.

I'm not sure these things are worthy of keeping in one's brain :-) The only reason why I remember it is that if you need to split UTF-8 encoded strings on US-ASCII delimiters it's perfectly possible to simply work at the byte level.